google/cloud/spanner_v1/database.py

```diff
@@ -880,17 +893,26 @@ def run_in_transaction(self, func, *args, **kw):
         # Sanity check: Is there a transaction already running?
         # If there is, then raise a red flag. Otherwise, mark that this one
         # is running.
-        if getattr(self._local, "transaction_running", False):
-            raise RuntimeError("Spanner does not support nested transactions.")
-        self._local.transaction_running = True
-
-        # Check out a session and run the function in a transaction; once
-        # done, flip the sanity check bit back.
-        try:
-            with SessionCheckout(self._pool) as session:
-                return session.run_in_transaction(func, *args, **kw)
-        finally:
-            self._local.transaction_running = False
+        with SessionCheckout(self._pool) as session:
+            observability_options = getattr(self, "observability_options", None)
+            with trace_call(
+                "CloudSpanner.Database.run_in_transaction",
+                session,
+                observability_options=observability_options,
+            ):
+                # Sanity check: Is there a transaction already running?
+                # If there is, then raise a red flag. Otherwise, mark that this one
+                # is running.
+                if getattr(self._local, "transaction_running", False):
+                    raise RuntimeError("Spanner does not support nested transactions.")
+                self._local.transaction_running = True
+
+                # Check out a session and run the function in a transaction; once
+                # done, flip the sanity check bit back.
+                try:
```

Comment:
The try block should start before with SessionCheckout since that can also throw error